### DISCUSSION TITLE
Clocked System on microsecond timescale

### DISCUSSION THREAD

**Discussion by vreinhd** [OpenModelica/OpenModelica#9562] [Q&A] [answered]
Dear all, 

I am pretty new to OpenModelica and interested in clocked simulations on the (sub)microsecond timescale and more.
I found that somehow the evaluation of clocks (even of PeriodicExactClocks) seems to be not exactly time-synchronized on these time-scales. The attached zip-file contains two Modelica Models which basically perform a comparison of different clocked signals.

![image](https://user-images.githubusercontent.com/116268682/196925178-20b1ca35-bb1c-4575-83c2-b71559bb3a98.png)

The uppermost system compares signals sampled with real clocks, the middle one signals sampled using exact clocks and the lower one signals with a true subsampling. Based on the input of the ramp I expected to get a step function with values 0, -0.0001, -0.0002 and -0.0003. On the microsecond scale this is true when simulating the system for a second - but after that I get sharp peaks at the instant the sampled signals should be synchrounous again.
This holds for both the upper and the middle system, only the true subsampling yields acceptable results. 
Simulating the same system with clocks of the period 1s and 4s yields correct results for all three systems.

Please excuse if my explanation is lacking details. I'll be happy to answer any question related to that problem and appreciate any comments!

[Clocktest_models.zip](https://github.com/OpenModelica/OpenModelica/files/9828933/Clocktest_models.zip)

![microsecond_clocks](https://user-images.githubusercontent.com/116268682/196923182-359cbeaf-f948-463c-b88b-e7c0720ccdec.png)
![second_clocks](https://user-images.githubusercontent.com/116268682/196923187-f85c71ac-d91c-41f8-a355-58be1f037e4f.png)

**Comment by casella** [comment, marked as answer]
@vreinhd, in the first case the two clocks are defined independently with two Real periods, so they are definitely not synchronous. Hence, there can be some drift between them, and I'm definitely not surprised that you get glitches. In fact, the whole clocked variables and synchronous semantics was invented exactly to overcome this problem, but if you use two independently defined clocks, you lose the synchronization properties.

The second example is a bit subtler, because the two clocks are defined as multiples of a base resolution clock. But also in this case, they are defined independently, so there is no guarantee of synchronism when you compare them.

Only in the third case the two clocked partitions are related by a sub-clock partitioning relationship, see [Section 16.7.4](https://specification.modelica.org/maint/3.5/synchronous-language-elements.html#sub-clock-partitioning) of the language specification. This makes sure that the two clocks tick together when it's time to do so.

Bottom line, if you want clocks with rational relationships between their periods to always meet when it's time, make sure you derive them from _the same base clock_, via super- or sub-sampling. If you do so, you have formal theoretical guarantees that things will work properly. That's the whole point of using synchronous language semantics 😃

**Comment by vreinhd** [comment]
@casella, thank you very much for the explanation.

Just two short remarks: 

1. Users used to work with Simulink seem to find the sub-sampling semantics a bit inconvenient for systems with multipe switching frequencies. I will probably try to build a generic "multiclock" for that, based on a common base clock. 

2. The documentation of Modelica.Clocked.ClockSignals.Clocks.PeriodicExactClock contains "All clocks with rational number definitions are exactly time synchronized to each other." After your explanation this seems to be a bit misleading to me. Maybe a remark like "For exact synchronism use sub-clocking" in the documentation could help?

**Reply by casella** [reply]
>  1. Users used to work with Simulink seem to find the sub-sampling semantics a bit inconvenient for systems with multiple switching frequencies. I will probably try to build a generic "multiclock" for that, based on a common base clock.

I remember some discussions with the French guys who invented synchronous semantics, that were stressing the fact that the foundations of Simulink clocked systems are not very rigorous. Of course rigour always comes with a price 😅. However, it is likely that periodic exact clocks are what you really need to avoid this inconvenience, see below.

>  2. The documentation of Modelica.Clocked.ClockSignals.Clocks.PeriodicExactClock contains "All clocks with rational number definitions are exactly time synchronized to each other." After your explanation this seems to be a bit misleading to me. Maybe a remark like "For exact synchronism use sub-clocking" in the documentation could help?

@vreinhd, thanks for actually raising this issue. I admit I was not 100% sure of what I wrote about the second case. Intuitively I also thought that the whole point of those clock generators was to have implicitly synchronized clocks, but at a first glance, I was not able to map that on the language specification text. However, on a second reading I noticed the following statement in [Section 16.3](https://specification.modelica.org/maint/3.5/synchronous-language-elements.html#clock-constructors):
> one of the following mutually exclusive associations of clocks are possible in one base partition:
> 1. One or more rational interval clocks, provided they are consistent with each other, see [section 16.7.5](https://specification.modelica.org/maint/3.5/synchronous-language-elements.html#sub-clock-inferencing).

This means that if you have two compatible rational clocks (like the ones contained in PeriodicExactClock), they _can_ live in the same base clock partition, i.e. they can be synchronized. Per se this doesn't mean they _should_, but I'm not 100% sure. It could also be that the OMC implementation is not correct. 

Let me get an even simpler MWE to sort out the issue:
```modelica
model SameClock
  Clock Clock1(1, 1000);
  Clock Clock2(1,1000);
  Real x1(start = 0, fixed = true);
  Real x2(start = 0, fixed = true);
equation
  when Clock1 then
    x1 = previous(x1) + 1;
  end when;
  when Clock2 then
    x2 = previous(x2) + 1;
 end when;
end SameClock;
```
Q: are `x1` and `x2` in the same clocked partition (hence synchronized) or not?

By compiling this MWE with -d=optdaedump one gets:
```
######################################## 
pre-optimization module clockPartitioning (simulation) 
######################################## 
 
 
clocked partition(1) 
======================================== 
 
Variables (1) 
======================================== 
1: x1:VARIABLE(start = 0.0 fixed = true ) SameClock type: Real 
 
 
Equations (1, 1) 
======================================== 
1/1 (1): x1 = 1.0 + previous(x1)   [dynamic |0|0|0|0|] 
 
clocked partition(2) 
======================================== 
 
Variables (1) 
======================================== 
1: x2:VARIABLE(start = 0.0 fixed = true ) SameClock type: Real 
 
Equations (1, 1) 
======================================== 
1/1 (1): x2 = 1.0 + previous(x2)   [dynamic |0|0|0|0|] 

```
So, OMC currently thinks these are two different partitions, hence the behaviour you observed.

It is not clear to me if merging them into the same base clock partition (in fact in the same clock partition in this case, since there is no sub-supersampling here) is mandatory or just a tool optimization, since they are consistent with each other. I just opened modelica/ModelicaSpecification#3265 to get some clarification from the working group of the Modelica Association that defines the language spec, please follow the discussion there. Once it is clear what is meant by the language specification, we will implement that.

Keeping @AnHeuermann, @kabdelhak, and @phannebohm in the loop.
